### PR TITLE
Offline adjustments

### PR DESCRIPTION
**Pull request by cymbaum**
Adds Offline Conversion Adjustments functionality

### CODE REVIEW

File: megalista_dataflow/data_sources/big_query/big_query_data_source.py [hunk 1]
```diff
@@ -173,13 +184,17 @@ def _get_schema_fields(self):
             return SchemaField("uuid", "string"), SchemaField("timestamp", "timestamp")
         if self._transactional_type == TransactionalType.GCLID_TIME:
             return SchemaField("gclid", "string"), SchemaField("time", "string"), SchemaField("timestamp", "timestamp")
+        if self._transactional_type == TransactionalType.ORDER_ID:
+            return SchemaField("order_id", "string"), SchemaField("timestamp", "timestamp")
```
Comment (Mr-Lopes) [inline megalista_dataflow/data_sources/big_query/big_query_data_source.py:188]: Add time column if necessary (comments above)

**Review by Mr-Lopes — CHANGES_REQUESTED** [review]
Thank you for this! I've added a few comments and I also noticed that two unit tests are broken. Could you check?

File: megalista_dataflow/data_sources/big_query/big_query_data_source.py [hunk 2]
```diff
@@ -148,7 +158,8 @@ def write_transactional_info(self, rows, execution: Execution):
                 bq_rows = all_bq_rows[i: i + _BIGQUERY_PAGE_SIZE]
                 partial_results.append(client.insert_rows(table,
                     bq_rows,
-                    self._get_schema_fields()))
+                    self._get_schema_fields(),
+                    retry=Retry(deadline=120)))
```
[the remark this answers is not in the document]
Comment (cymbaum) [reply megalista_dataflow/data_sources/big_query/big_query_data_source.py:162]: do you mean (utils.py#L96)? 
I believe it is not called here in the insert_rows from the BQ client. This retry is preventing some transient errors described here: https://github.com/googleapis/python-bigquery/issues/1331
The retry + backoff seemed to resolve the issue.

File: megalista_dataflow/data_sources/big_query/big_query_data_source.py [hunk 3]
```diff
@@ -92,6 +93,10 @@ def _retrieve_data_transactional(self, executions: ExecutionsGroupedBySource) ->
                 template = "SELECT $query_cols FROM `$table_name` AS data \
                                 LEFT JOIN `$uploaded_table_name` AS uploaded USING(gclid, time) \
                                 WHERE uploaded.gclid IS NULL;"
+            elif self._transactional_type == TransactionalType.ORDER_ID:
+                template = "SELECT $query_cols FROM `$table_name` AS data \
+                                LEFT JOIN `$uploaded_table_name` AS uploaded USING(order_id) \
```
[the remark this answers is not in the document]
Comment (cymbaum) [reply megalista_dataflow/data_sources/big_query/big_query_data_source.py]: Done.

File: megalista_dataflow/data_sources/big_query/big_query_data_source.py [hunk 4]
```diff
@@ -121,6 +126,12 @@ def _ensure_control_table_exists(self, client: Client, uploaded_table_name: str)
                             time STRING OPTIONS(description= 'Original time')) \
                             PARTITION BY _PARTITIONDATE \
                             OPTIONS(partition_expiration_days=15)"
+        elif self._transactional_type == TransactionalType.ORDER_ID:
+            template = "CREATE TABLE IF NOT EXISTS `$uploaded_table_name` ( \
+                            timestamp TIMESTAMP OPTIONS(description= 'Event timestamp'), \
+                            order_id STRING OPTIONS(description= 'Order Id (transaction Id)')) \
```
[the remark this answers is not in the document]
Comment (cymbaum) [reply megalista_dataflow/data_sources/big_query/big_query_data_source.py]: Done.

File: megalista_dataflow/data_sources/big_query/big_query_data_source.py (hunk 1 above)
Comment (cymbaum) [reply megalista_dataflow/data_sources/big_query/big_query_data_source.py]: Done.

File: megalista_dataflow/data_sources/big_query/big_query_data_source.py [hunk 5]
```diff
@@ -173,13 +184,17 @@ def _get_schema_fields(self):
...
+            return SchemaField("order_id", "string"), SchemaField("timestamp", "timestamp")
         raise Exception(f'Unrecognized TransactionalType: {self._transactional_type}. Source="{self._source_name}". Destination="{self._destination_name}"')
 
     def _get_bq_rows(self, rows, now):
         if self._transactional_type == TransactionalType.UUID:
             return [{'uuid': row['uuid'], 'timestamp': now} for row in rows]
         if self._transactional_type == TransactionalType.GCLID_TIME:
             return [{'gclid': row['gclid'], 'time': row['time'], 'timestamp': now} for row in rows]
+        if self._transactional_type == TransactionalType.ORDER_ID:
+            return [{'order_id': row['order_id'], 'timestamp': now} for row in rows]
```
[the remark this answers is not in the document]
Comment (cymbaum) [reply megalista_dataflow/data_sources/big_query/big_query_data_source.py]: Done.

File: megalista_dataflow/data_sources/data_schemas.py [hunk 6]
```diff
@@ -52,6 +52,23 @@
         ],
         'groups': []
     },
+    'ADS_OFFLINE_CONVERSION_ADJUSTMENT_GCLID': {
+        'columns': [
+            {'name': 'gclid', 'required': True, 'data_type': 'string'},
+            {'name': 'time', 'required': True, 'data_type': 'string'},
+            {'name': 'conversion_time', 'required': True, 'data_type': 'string'},
+            {'name': 'amount', 'required': False, 'data_type': 'string'}
```
[the remark this answers is not in the document]
Comment (cymbaum) [reply megalista_dataflow/data_sources/data_schemas.py:60]: Done.

File: megalista_dataflow/data_sources/data_schemas.py [hunk 7]
```diff
@@ -52,6 +52,23 @@
...
+        ],
+        'groups': []
+    },
+    'ADS_OFFLINE_CONVERSION_ADJUSTMENT_ORDER_ID': {
+        'columns': [
+            {'name': 'order_id', 'required': True, 'data_type': 'string'},
+            {'name': 'time', 'required': True, 'data_type': 'string'},
+            {'name': 'amount', 'required': False, 'data_type': 'string'}
```
[the remark this answers is not in the document]
Comment (cymbaum) [reply megalista_dataflow/data_sources/data_schemas.py:68]: Done.

File: megalista_dataflow/uploaders/google_ads/conversions/google_ads_offline_conversion_adjustments_uploader.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            raise ValueError(
+                'Missing Metadata 1. Received {}'.format(str(destination))
+            )
+        
+        if not destination[2]:
+            raise ValueError(
+                'Missing Metadata 3. Received {}'.format(str(destination))
+            )
+
+    @utils.safe_process(logger=logging.getLogger('megalista.GoogleAdsOfflineUploader'))
```
[the remark this answers is not in the document]
Comment (cymbaum) [reply megalista_dataflow/uploaders/google_ads/conversions/google_ads_offline_conversion_adjustments_uploader.py]: Done. Thanks :)

**Review by Mr-Lopes — CHANGES_REQUESTED** [review]
Please see comments

**Review by Mr-Lopes — APPROVED** [review]
LGTM

**Pull request merged**